First pass on proxy peering docs (#15790)
gravitational/teleport: docs/pages/architecture/proxy-peering.mdx

---
title: Proxy Peering
description: How Teleport implements more efficient networking with proxy peering.
---

<Details
  title="Version warning"
  opened={true}
  scope={["oss", "enterprise"]}
  scopeOnly={true}
  min="10.0"
>
  Proxy peering is available starting from Teleport `10.0`.
</Details>

> ## Overview

Reviewer: I think we can remove this, since it's implied that the first paragraph is an overview.

Proxy peering is a type of tunnel strategy configurable in Teleport. A tunnel
strategy defines the behavior for agent reverse tunnel connections as well as

> how client are connected to those reverse tunnels. A Teleport Cluster's tunnel

Reviewer: ```suggestion
how client are connected to those reverse tunnels. A Teleport cluster's tunnel
```

strategy is configurable in the Teleport Auth service.

> Proxy peering enables an agent to be reachable without it connecting to every

Reviewer: - This seems like the main point of Proxy Peering—we should put this paragraph first.

- What is an "agent" in this context? I'm guessing its an instance of the Database Service, SSH Service, etc., but will a reader know that? (This applies to the term "agents" throughout this page.)

Author: Should we define agent in this page or swap it for `Teleport resource service` and provide the list of resources this refers to?

> proxy instance. This allows proxies to scale horizontally without increasing

Reviewer: ```suggestion
Proxy Service instance. This allows Proxy Service instances to scale horizontally without increasing
```
The term "proxy" comes up in various contexts in our docs, so we should be more precise here so readers can understand our architecture more easily.

the number of connections created by agents.

## How it works

### Proxy Service

> A GRPC service on Teleport Proxy servers provides an API for establishing a

Reviewer: ```suggestion
A GRPC service on each Teleport Proxy Service instance provides an API for establishing a
```

> bi-directional connection to the agents connected to that proxy. Proxy servers

Reviewer: ```suggestion
bi-directional connection to the agents connected to that instance.
```

manage a GRPC client to all other proxies in the cluster.

> Routing information on which proxies each agent is connected to is stored in

Reviewer: Same comment re: "proxies" and "agent"

Teleport's backend and propagated to each proxy instance.

The routing information and GRPC service allow a Proxy to identify where an

> agent is connected and create an end to end connection from a client to that

Reviewer: ```suggestion
agent is connected and create an end-to-end connection from a client to that
```

agent.

> ### Reverse Tunnel Agents

Reviewer: ```suggestion
### Reverse tunnel agents
```
Headings use sentence case

> Normally teleport agents need to create a reverse tunnel to every Teleport Proxy

Reviewer: ```suggestion
Normally Teleport agents need to create a reverse tunnel to every Teleport Proxy
```

to ensure a client is able to reach the agent. With proxy peering this is no
longer a requirement. When proxy peering is enabled agents will automatically
change their behavior to connect to the configured number of proxy servers.

> By default agents are configured to connect to a single proxy server. For high

Reviewer: This paragraph seems like it should be much higher in the page, e.g., in the first section. Otherwise, there's not a lot of context for why a reader would want to use Proxy Peering.

Author: Moved to top section. [0969b49](https://github.com/gravitational/teleport/pull/15790/commits/0969b499452386e8d67bd97cc58b3c74fd092176)

availability a cluster administrator may configure agents to connect to 2 or
more proxy servers.

> ![Teleport Proxy Peering](../../img/architecture/proxy-peering@1.2x.svg)

Reviewer: This image does not render. Did you view this page on the local dev server?

Author: Hmm, it works for me locally :thinking:

Author: Try pulling down the latest changes and see if it works. You may have missed the commit that adds the diagram.

## Next Steps

> - See the [migration guide](../management/operations/proxy-peering.mdx) to learn how to upgrade an existing cluster to use

Reviewer: The link does not render

Author: This and the svg both work for me as expected running `yarn dev` locally.

---
title: Proxy Peering
description: How Teleport implements more efficient networking with Proxy Peering.
---

<Details
  title="Version warning"
  opened={true}
  scope={["oss", "enterprise"]}
  scopeOnly={true}
  min="10.0"
>
  Proxy Peering is available starting from Teleport `10.0`.
</Details>

Proxy Peering enables an agent to be reachable without it connecting to every
Teleport Proxy instance. This allows Teleport Proxy instances to scale horizontally
without increasing the number of connections created by agents.

> Proxy Peering is a type of tunnel strategy configurable in Teleport. A tunnel

Reviewer: I think we can remove this paragraph. The next paragraph explains how Proxy Peering works in relation to the Proxy Service's default behavior, and the previous paragraph explains the benefit.

  A Teleport agent is a Teleport instance that provides access to resources in your infrastructure, i.e., by running the SSH,

>   Kubernetes, Database, Application, and Desktop Services.

Reviewer 2: ```suggestion
  Kubernetes, Database, Application, or Desktop Services.
```

---
title: Proxy Peering
description: How Teleport implements more efficient networking with Proxy Peering.
---

<Details
  title="Version warning"
  opened={true}
  scope={["oss", "enterprise"]}
  scopeOnly={true}
  min="10.0"
>
  Proxy Peering is available starting from Teleport `10.0`.
</Details>

Proxy Peering enables Teleport agents to be reachable without connecting to
every Teleport Proxy Service. This allows Teleport Proxy instances to scale
horizontally without increasing the number of connections created by agents.

<Admonition title="Note" type="tip">

>   A Teleport agent is a service that provides access to a Teleport service such as SSH,

Reviewer: ```suggestion
  A Teleport agent is a Teleport instance that provides access to resources in your infrastructure, i.e., by running the SSH,
```

>   Kubernetes, Database, Application, and Desktop services.

Reviewer: ```suggestion
  Kubernetes, Database, Application, and Desktop Services.
```
Since we capitalize "Desktop Service", "Application Service", etc.

</Admonition>

By default, Teleport agents need to create a reverse tunnel to every Teleport Proxy
to ensure a client is able to reach the agent. With Proxy Peering this is no
longer a requirement. When Proxy Peering is enabled agents will automatically
change their behavior to connect to the configured number of Teleport Proxy
instances.

## How it works

### Proxy Service

> A gRPC service on each Teleport Proxy instance provides an API for establishing a

Reviewer: ```suggestion
A gRPC service on each Teleport Proxy Service instance provides an API for establishing a
```

bi-directional connection to the agents connected to that Teleport Proxy. Teleport
Proxy instances manage a gRPC client to all other Teleport Proxy instances in the
cluster.

Routing information on which Teleport Proxy instances each agent is connected to
is stored in Teleport's backend and propagated to each Teleport Proxy instance.

The routing information and gRPC service allow a Teleport Proxy to identify which
Proxy Service instance an agent is connected to and create an end-to-end connection
from a client to that agent. This allows for access to the agent without connecting
to the same Teleport Proxy instance initially.

### Reverse tunnel agents

Agents will check whether you have enabled Proxy Peering before

> attempting to create a reverse tunnel to a Teleport Proxy instance. This

Reviewer: I think we should delete the last sentence of this paragraph ("This allows an agent to dynamically adjust its behavior accordingly"), since it doesn't provide any information we don't already get in the next paragraph.
